### PR TITLE
#908 Quality of Service - Configuration

### LINKED ISSUE #908: Quality of Service - Configuration
## Background
- As per the [Quality of Service](https://ocelot.readthedocs.io/en/latest/features/qualityofservice.html), documentation we can set up `TimeoutValue` in isolation of the `ExceptionsAllowedBeforeBreaking` and `DurationOfBreak` options. I'm trying to only use `TimeoutValue` property without the other two.
 
## Expected Behavior
- Expected Behavior is that if I have just configure `TimeoutValue` then I should be able to call the API Gateway without any issue.

## Actual Behavior
- Configuring just the `TimeoutValue` returns `500 Internal Server Error`. Here are logs:
`````````````````````````
[WRN] requestId: 80000010-0007-ff00-b63f-84710c7967bb, previousRequestId: no previous request id, message: Error making http request, exception: System.ArgumentException: The enumerable of policies to form the wrap must contain at least two policies.
Parameter name: policies
   at Polly.Policy.WrapAsync(IAsyncPolicy[] policies) in C:\projects\polly\src\Polly.Shared\Wrap\PolicyWrapSyntaxAsync.cs:line 93
   at Ocelot.Provider.Polly.PollyCircuitBreakingDelegatingHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.HttpClient.FinishSendAsyncBuffered(Task`1 sendTask, HttpRequestMessage request, CancellationTokenSource cts, Boolean disposeCts)
   at Ocelot.Requester.HttpClientHttpRequester.GetResponse(DownstreamContext context)

[WRN] requestId: 80000010-0007-ff00-b63f-84710c7967bb, previousRequestId: no previous request id, message: Error Code: UnableToCompleteRequestError Message: Error making http request, exception: System.ArgumentException: The enumerable of policies to form the wrap must contain at least two policies.
Parameter name: policies
   at Polly.Policy.WrapAsync(IAsyncPolicy[] policies) in C:\projects\polly\src\Polly.Shared\Wrap\PolicyWrapSyntaxAsync.cs:line 93
   at Ocelot.Provider.Polly.PollyCircuitBreakingDelegatingHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.HttpClient.FinishSendAsyncBuffered(Task`1 sendTask, HttpRequestMessage request, CancellationTokenSource cts, Boolean disposeCts)
   at Ocelot.Requester.HttpClientHttpRequester.GetResponse(DownstreamContext context) errors found in ResponderMiddleware. Setting error response for request path:/api/HomeAssessment/Codification/slowService/20, request method: GET
2019-05-31 11:54:27.395 +10:00 [ERR] <End Request> GET /api/HomeAssessment/Codification/slowService/20 500 8ms
`````````````````````````

## Additional Information
This issue can be also be reproduced with `Ocelot.AcceptanceTests` -> `PollyQoSTests`. If we remove property  `ExceptionsAllowedBeforeBreaking` from `should_timeout` and `should_not_timeout` methods, we get `InternalServerError`. As per documentation, this should not be case.

### PR DESCRIPTION
**Pull request by DanHarltey**
## Fixes #908 
- #908 

Fixing an issue where only setting timeout value in QoS would result in a exception getting thrown.

## Proposed Changes
  - Added small check to see if there is only Policy
  - Update a unit test to trigger the behavior

### CODE REVIEW

**Review by eddex — DISMISSED** [review]
lgtm

**Comment by somasundar1997** [comment]
Hi when can I get this release.

**Comment by somasundar1997** [comment]
@eddex when I can expect this to get merged with the main branch

**Comment by eddex** [comment]
> @eddex when I can expect this to get merged with the main branch

I'm not a maintainer of Ocelot, just a user that wants to get this fixed too. I think @raman-m might be able to merge it.

**Comment by raman-m** [comment]
@DanHarltey  Hi Dan!

Requesting to **master** is wrong!
You can request PRs to **develop** branch only!

Please, change the base (target) branch!
I would recommend to pull latest commits from **base:develop** and please apply your changes on the top commit!

Pay attention that CI pipelines are working for **main** and **develop** branches only!
I have to check CI pipeline status also before starting code review...
Also, I have no right to change PR target/base branch. So, only PR author can edit the PR!

**Review by eddex — APPROVED** [review]

**Review by eddex — DISMISSED** [review]

**Review by raman-m — DISMISSED** [review]
```diff
+ Approved
```

**Comment by raman-m** [comment]
@eddex  Hi Marco!
Sorry for stale review!

Could you review old and my changes once again please?

Please note that real bug fix is in **try**-block of the `SendAsync` method of the `PollyCircuitBreakingDelegatingHandler` class: [Lines 28-41](https://github.com/DanHarltey/Ocelot/blob/master/src/Ocelot.Provider.Polly/PollyCircuitBreakingDelegatingHandler.cs#L28-L41)

And the core of the bug fix is in lines [L33-L35](https://github.com/DanHarltey/Ocelot/blob/master/src/Ocelot.Provider.Polly/PollyCircuitBreakingDelegatingHandler.cs#L33-L35) which was proposed by @DanHarltey in his solution from initial commit [91eed16d96cf13e97d6215bcfb48f1ddf5d0fadd](https://github.com/DanHarltey/Ocelot/commit/91eed16d96cf13e97d6215bcfb48f1ddf5d0fadd).
So, the `Policy.WrapAsync(policies)` generates exception if collection has less than 2 elements (**0** and **1**). The fix takes this conditions into account.
Finally, we can apply wrapping operation for policies collections with 2 elements or more. 

Also, I have added unit tests as the `PollyCircuitBreakingDelegatingHandlerTests` class to cover the cases above.

I would appreciate your code review!

File: src/Ocelot.Provider.Polly/PollyCircuitBreakingDelegatingHandler.cs
Comment (eddex) [inline src/Ocelot.Provider.Polly/PollyCircuitBreakingDelegatingHandler.cs:35]: I would reverse this check to reduce cognitive complexity. First check for 0 with `policies.Any()`, then check for 1 and otherwise wrap the policies because there are more.

```suggestion
                    IAsyncPolicy policy = policies.Length == 1
                        ? policies[0]
                        : Policy.WrapAsync(policies);
```

**Review by eddex — DISMISSED** [review]

File: src/Ocelot.Provider.Polly/PollyCircuitBreakingDelegatingHandler.cs
Comment (raman-m) [reply src/Ocelot.Provider.Polly/PollyCircuitBreakingDelegatingHandler.cs:35]: Good point! But... We have a deal with white box, not black box.
Your point about "cognitive complexity" is right for black boxes only.
So, we can review and even change source code of the feature, because we have the white box.
I am saying about concrete implementation, about the [PollyQoSProvider](../../blob/develop/src/Ocelot.Provider.Polly/PollyQoSProvider.cs#L14) class.
This class is injected to delegating handler in the `AddPolly` extension method, see [line 38](../../blob/develop/src/Ocelot.Provider.Polly/OcelotBuilderExtensions.cs#L38) plz.

Let's return back to `PollyQoSProvider` [line 57](../../blob/develop/src/Ocelot.Provider.Polly/PollyQoSProvider.cs#L57) and we will see that `CircuitBreaker` object is initialized with 2 policies.
**Finally**, we have two cases:
- There is no QoS feature enabled. **0** policies
- QoS feature is enabled and initialized by **2** policies

So, **0** or **2** policies.
I believe my code flow is correct. We check for **0** policies first by `Any`.
If the feature enabled we check for **2** policies first.
But in theory one policy can be defined, and this is the last operator in code flow.

Truly speaking, Tom forgot to wrap these 2 policies while creating circuit breaker in the `PollyQoSProvider` class. That's why we have this bug issue.
Finally, I would prefer to move wrapping logic to the `PollyQoSProvider` class. Let's see how Tom will review this PR.

Comment (eddex) [reply src/Ocelot.Provider.Polly/PollyCircuitBreakingDelegatingHandler.cs:35]: The order doesn't really matter. It would also be an improvement to just check for `> 1` instead of `>= 2`. I think this is more readable because _"more than 1"_ is more straight forward than _"more than 2 or exactly 2"_.

```suggestion
                    IAsyncPolicy policy = policies.Length > 1
                        ? Policy.WrapAsync(policies)
                        : policies[0];
```

Comment (raman-m) [reply src/Ocelot.Provider.Polly/PollyCircuitBreakingDelegatingHandler.cs]: Fixed in commit [Fix code review issue: "more than 1"](https://github.com/ThreeMammals/Ocelot/pull/1279/commits/7b431bc8b8fc871cc1356cea8eeda9740da7adcd)

**Review by raman-m — APPROVED** [review]
```diff
+ Approved
```
@DanHarltey Dan, thanks for the nice fix!
@eddex  Marco, thanks for review and collaboration!

**Comment by raman-m** [comment]
@TomPallister  Tom, don't be confused by [DanHarltey:master](https://github.com/DanHarltey/Ocelot/commits/master) source branch!
Dan's **master** branch has the latest top-commits from **base**. This **master** branch was updated by **develop** commits during **rebase** operation. See the logs above plz!

Could you approve this PR and merge it please?

**Review by raman-m — APPROVED** [review]

**Comment by raman-m** [comment]
@DanHarltey Hi Dan!
I don't see **develop** branch in [your fork](https://github.com/DanHarltey/Ocelot/branches)! Your fork is too old.

Could you [Sync fork](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/syncing-a-fork) please? So, new **develop** branch will occur with all top commits!

---

If [Sync fork](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/syncing-a-fork) button won't help to make develop branch as default one.
Run this script please:
```shell
git remote add upstream https://github.com/ThreeMammals/Ocelot.git // add original repo as upstream remote
git fetch upstream                                                 // pull changes from original repo
git checkout -b develop upstream/develop                           // pull down develop branch locally
git push -u origin develop                                         // push develop branch to fork
```
Don't forget to make **develop** as default branch in GitHub.
Let me know updating results please!

**Comment by DanHarltey** [comment]
Hi @raman-m,
Thank you for your instructions. I needed them. This has now been successfully completed.

**Review by raman-m — DISMISSED** [review]

**Comment by raman-m** [comment]
> @DanHarltey commented [on July 21](#issuecomment-1646200685)

Dan, JFYI
Because we have synced **develop** branch, I've successfully rebased [feature branch](/DanHarltey/Ocelot/tree/master) onto [ThreeMammals:develop](../tree/develop).

Thanks for updating of **develop** branch!

File: src/Ocelot.Provider.Polly/PollyCircuitBreakingDelegatingHandler.cs
Comment (RaynaldM) [inline src/Ocelot.Provider.Polly/PollyCircuitBreakingDelegatingHandler.cs:30]:
```suggestion
                if (!policies.Any())  return await base.SendAsync(request, cancellationToken);
```

File: src/Ocelot.Provider.Polly/PollyCircuitBreakingDelegatingHandler.cs
Comment (RaynaldM) [inline src/Ocelot.Provider.Polly/PollyCircuitBreakingDelegatingHandler.cs:39]:
```suggestion
              
```

**Review by RaynaldM — APPROVED** [review]

**Review by raman-m — APPROVED** [review]

**Review by RaynaldM — APPROVED** [review]

**Pull request merged**
